File: 2429-design-a-food-rating-system/2429-design-a-food-rating-system.py
class FoodRatings:

    def __init__(self, foods: List[str], cuisines: List[str], ratings: List[int]):
        self.foodMap = {}
        self.cuisineHighest = {}
        for i in range(len(foods)):
            self.foodMap[foods[i]] = (cuisines[i], ratings[i])
            if cuisines[i] not in self.cuisineHighest:
                self.cuisineHighest[cuisines[i]] = SortedSet()
            self.cuisineHighest[cuisines[i]].add((-ratings[i], foods[i]))

    def changeRating(self, food: str, newRating: int) -> None:
        self.cuisineHighest[self.foodMap[food][0]].remove((-self.foodMap[food][1], food))
        self.cuisineHighest[self.foodMap[food][0]].add((-newRating, food))
        self.foodMap[food] = (self.foodMap[food][0], newRating)


    def highestRated(self, cuisine: str) -> str:
        return self.cuisineHighest[cuisine][0][1]


# Your FoodRatings object will be instantiated and called as such:
# obj = FoodRatings(foods, cuisines, ratings)
# obj.changeRating(food,newRating)
# param_2 = obj.highestRated(cuisine)

File: 2429-design-a-food-rating-system/2429-design-a-food-rating-system.cpp
class FoodRatings {
public:
    FoodRatings(vector<string>& foods, vector<string>& cuisines, vector<int>& ratings) {
        for(int i = 0; i < foods.size(); i++)
        {
            foodMap[foods[i]] = {cuisines[i], ratings[i]};
            cuisineHighest[cuisines[i]].insert({-ratings[i], foods[i]});
        }
    }
    
    void changeRating(string food, int newRating) {
        cuisineHighest[foodMap[food].first].erase({-foodMap[food].second, food});
        cuisineHighest[foodMap[food].first].insert({-newRating, food});
        foodMap[food].second = newRating;
    }
    
    string highestRated(string cuisine) {
        return (*cuisineHighest[cuisine].begin()).second;
    }
private:
    unordered_map<string, pair<string, int>> foodMap;
    unordered_map<string, set<pair<int, string>>> cuisineHighest;
};

/**
 * Your FoodRatings object will be instantiated and called as such:
 * FoodRatings* obj = new FoodRatings(foods, cuisines, ratings);
 * obj->changeRating(food,newRating);
 * string param_2 = obj->highestRated(cuisine);
 */